# Datan analysointi K2021 - Tehtävät kerta 4

Palauta ratkaisusi Moodleen, kun olet ensin testannut ratkaisusi toimivuuden Notebookilla tai omalla koneellasi. Kun olet palauttanut tehtäväsi Moodleen, ne arvioidaan pistein:
* 1,0 p. Vastaus oikein ensimmäisellä yrityksellä.
* 0,5 p. Vastaus oikein toisella yrityksellä.
* 0 p. Vastaus on annettu kolmannella yrityksellä.

**Tee tehtävät ilman for-silmukoita tms.**

Euroopan ympäristökeskus (EEA) julkaisee dataa uusista rekisteröidyistä autoista, mitä käytetään tämän harjoituksen tehtävissä. Jos olet kiinnostunut tarkemmin datasta, voit käydä tutustumassa siihen EEA:n [lataussivulta](https://www.eea.europa.eu/ds_resolveuid/53127e6b74ab4d28a12d4b365260bc38).

### 1. Autodatan yhdistäminen

Tee funktio autot(url1, url2), joka tekee järjestyksessä seuraavaa:
* lataa datat DataFrameihin url1:sta ja url2:sta, joissa erotinmerkkinä on puolipiste (';'),
* yhdistää uuteen DataFrameen autoihin liittyvät datat url1:stä ja url2:sta (tutki tarvittaessa mitä dataa urleista on saatavilla),
* tekee juuri yhdistetystä DataFramesta Serieksen, josta käy ilmi kuinka monta automerkkiä datassa on (automerkki on sarakkeessa 'Cn'),
* tulostaa em. Serieksestä viisi eniten esiintyvää automerkkiä.

Tässä tehtävässä:
* url1=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA1.csv
* url2=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA2.csv

Esimerkiksi, autot(url1, url2):

| | |
|:---|:---|
|OCTAVIA   | 396|
|FABIA     | 377|
|RAPID     | 197|
|SUPERB    | 159|
|KAROQ     | 148|
|Name: Cn, dtype: int64 | |

In [7]:
def autot(url1, url2):
    # Kirjoita vastauksesi tähän
    import pandas as pd    
    df1 = pd.read_csv(url1, 
                     engine='python', 
                     encoding='utf-8-sig', 
                     sep=';', 
                     )
    df2 = pd.read_csv(url2, 
                     engine='python', 
                     encoding='utf-8-sig', 
                     sep=';', 
                     )
    #print(df1)
    #print(df2)
    df3 = pd.concat([df1, df2], ignore_index = True)
    tulos = df3['Cn'].value_counts().head(5)
    
    # Älä kirjoita muita print komentoja ennen tulos muuttujan tulostusta.
    print(tulos)
    
# Testaa funktiotasi:
url1='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA1.csv'
url2='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA2.csv'

autot(url1, url2)

OCTAVIA    396
FABIA      377
RAPID      197
SUPERB     159
KAROQ      148
Name: Cn, dtype: int64


### 2. Autodatan yhdistäminen ja ryhmittely

Tee funktio autot(url_a1, url_a2, url_b1, url_b2), joka tekee järjestyksessä seuraavaa:
* lataa datat DataFrameihin osoitteista (url_a1, url_a2, url_b1, url_b2), joissa erotinmerkkinä on puolipiste (';'),
* yhdistää uuteen DataFrameen 'dfa' autoihin liittyvät datat url_a1:sta ja url_a2:sta (tutki tarvittaessa mitä dataa urleista on saatavilla),
* yhdistää uuteen DataFrameen 'dfb' autoihin liittyvät datat url_b1:sta ja url_b2:sta (tutki tarvittaessa mitä dataa urleista on saatavilla),
* yhdistää DataFramet 'dfa' ja 'dfb' yhdeksi DataFrameksi 'df' sarakkeen 'ID' perusteella,
* jos DataFramessa 'df' on sarakkeessa auton moottorin teho 'ep (KW)' nan arvoja, korvataan ne lukuarvolla nolla,
* tuottaa DataFramen 'tulos', johon on laskettu autojen keskimääräiset tehot sarakkeesta 'ep (KW)' ryhmiteltynä auton valmistajan mukaan 'Mh' (tutki tarvittaessa groupby [dokumentaatiosta](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html?highlight=groupby#pandas.DataFrame.groupby) kuinka groupby palauttaa indeksin),
* järjestää DataFramen 'tulos' rivit keskimääräisen tehon mukaan 'ep (KW)' suurimmasta pienimpään.
* tulostaa edellä jäjestetystä DataFramesta viisi ensimmäistä riviä.

Tässä tehtävässä:
* url_a1=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA1.csv
* url_a2=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA2.csv
* url_b1=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB1.csv
* url_b2=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB2.csv

Esimerkiksi, autot(url_a1, url_a2, url_b1, url_b2):

| | Mh  |   ep (KW)|
|:---|:---|:---|
|6 |           TESLA | 386.000000|
|0 |          AA-IVA | 267.750000|
|2 | FORD WERKE GMBH |  89.741007|
|5 |           SKODA |  81.372515|
|3 |             KIA |  77.000000|

In [45]:
def autot(url_a1, url_a2, url_b1, url_b2):
    # Kirjoita vastauksesi tähän
    import pandas as pd    
    df_a1 = pd.read_csv(url_a1, sep=';')
    df_a2 = pd.read_csv(url_a2, sep=';')
    df_b1 = pd.read_csv(url_b1, sep=';')
    df_b2 = pd.read_csv(url_b2, sep=';')    
        
    dfa = pd.concat([df_a1, df_a2], ignore_index = True)
    dfb = pd.concat([df_b1, df_b2], ignore_index = True)
        
    df = pd.merge(dfa, dfb, on = 'ID')
    
    df['ep (KW)'].fillna(0, inplace = True)    
    
    tulos = df.groupby(['Mh'], as_index=False)['ep (KW)'].mean()
    
    tulos = tulos.sort_values('ep (KW)', ascending = False)
    tulos = tulos.head(5)
    # Älä kirjoita muita print komentoja ennen tulos muuttujan tulostusta.
    print(tulos)
    
# Testaa funktiotasi:
url_a1='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA1.csv'
url_a2='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA2.csv'
url_b1='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB1.csv'
url_b2='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB2.csv'

autot(url_a1, url_a2, url_b1, url_b2)

                Mh     ep (KW)
6            TESLA  386.000000
0           AA-IVA  267.750000
2  FORD WERKE GMBH   89.741007
5            SKODA   81.372515
3              KIA   77.000000


### 3. Autodatan yhdistäminen ja ristiintaulukointi

Tee funktio autot(url_a1, url_a2, url_b1, url_b2), joka tekee järjestyksessä seuraavaa:
* lataa datat DataFrameihin osoitteista (url_a1, url_a2, url_b1, url_b2), joissa erotinmerkkinä on puolipiste (';'),
* yhdistää uuteen DataFrameen 'dfa' autoihin liittyvät datat url_a1:sta ja url_a2:sta (tutki tarvittaessa mitä dataa urleista on saatavilla),
* yhdistää uuteen DataFrameen 'dfb' autoihin liittyvät datat url_b1:sta ja url_b2:sta (tutki tarvittaessa mitä dataa urleista on saatavilla),
* yhdistää DataFramet 'dfa' ja 'dfb' yhdeksi DataFrameksi 'df' sarakkeen 'ID' perusteella,
* jos sarakkeessa 'Ft' on merkkijono 'DIESEL', korvataan se merkkijonolla 'Diesel',
* jos sarakkeessa 'Ft' on merkkijono 'PETROL', korvataan se merkkijonolla 'Petrol',
* tekee uuden DataFramen hyödyntäen [crosstab funktiota](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.crosstab.html) ja sopivaa numpy [funktiota](https://numpy.org/doc/stable/reference/), jossa on laskettu esimerkkituloksen mukaisesti keskimääräiset päästöt sarakkeesta 'Enedc (g/km)' autonvalmistaja kohtaisesti (sarake 'Mh') polttoainetyypeittäin (sarake 'Ft'),
* esittää edellä kuvatusta DataFramesta keskimääräiset päästöt yhden desimaalin tarkkuudella,
* valitsee uuteen DataFrameen sarakkeet 'Dielsel' ja 'Petrol',
* tulostaa edellä tuotetusta DataFramesta viisi ensimmäistä riviä.

Tässä tehtävässä:
* url_a1=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA1.csv
* url_a2=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA2.csv
* url_b1=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB1.csv
* url_b2=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB2.csv

Esimerkiksi, autot(url_a1, url_a2, url_b1, url_b2):

<table>
<thead>
  <tr>
    <th>Ft</th>
    <th>Diesel</th>
    <th>Petrol</th>
  </tr>
  <tr>
    <th>Mh</th>
    <th></th>
    <th></th>
  </tr>
</thead>
<tbody>
  <tr>
    <td>AA-IVA</td>
    <td>129.0</td>
    <td>194.7</td>
  </tr>
  <tr>
    <td>AUTOMOBILES CITROEN</td>
    <td>102.7</td>
    <td>105.6</td>
  </tr>
    <tr>
    <td>FORD WERKE GMBH</td>
    <td>112.9</td>
    <td>133.7</td>
  </tr>
    <tr>
    <td>KIA</td>
    <td>nan</td>
    <td>88.0</td>
  </tr>
    <tr>
    <td>NISSAN</td>
    <td>112.3</td>
    <td>127.7</td>
</tbody>
</table>


In [69]:
def autot(url_a1, url_a2, url_b1, url_b2):
    import pandas as pd
    import numpy as np
    df_a1 = pd.read_csv(url_a1, sep=';')
    df_a2 = pd.read_csv(url_a2, sep=';')
    df_b1 = pd.read_csv(url_b1, sep=';')
    df_b2 = pd.read_csv(url_b2, sep=';')    
        
    dfa = pd.concat([df_a1, df_a2], ignore_index = True)
    dfb = pd.concat([df_b1, df_b2], ignore_index = True)
            
    df = pd.merge(dfa, dfb, on = 'ID')
    
    df['Ft'] = df['Ft'].str.replace('DIESEL', 'Diesel')
    df['Ft'] = df['Ft'].str.replace('PETROL', 'Petrol')
    
    df2 = pd.crosstab(df['Mh'], df['Ft'], values = df['Enedc (g/km)'], aggfunc = np.mean).applymap("{:.1f}".format)
    df2 = df2[['Diesel', 'Petrol']]
    tulos = df2.head(5)
    print(tulos)
    
# Testaa funktiotasi:
url_a1='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA1.csv'
url_a2='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA2.csv'
url_b1='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB1.csv'
url_b2='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB2.csv'

autot(url_a1, url_a2, url_b1, url_b2)

Ft                  Diesel Petrol
Mh                               
AA-IVA               129.0  194.7
AUTOMOBILES CITROEN  102.7  105.6
FORD WERKE GMBH      112.9  133.7
KIA                    nan   88.0
NISSAN               112.3  127.7


### 4. Autodatan yhdistäminen ja ryhmittely monitasoiseksi DataFrameksi

Tee funktio autot(url_a1, url_a2, url_b1, url_b2), joka tekee järjestyksessä seuraavaa:
* lataa datat DataFrameihin osoitteista (url_a1, url_a2, url_b1, url_b2), joissa erotinmerkkinä on puolipiste (';'),
* yhdistää uuteen DataFrameen 'dfa' autoihin liittyvät datat url_a1:sta ja url_a2:sta (tutki tarvittaessa mitä dataa urleista on saatavilla),
* yhdistää uuteen DataFrameen 'dfb' autoihin liittyvät datat url_b1:sta ja url_b2:sta (tutki tarvittaessa mitä dataa urleista on saatavilla),
* yhdistää DataFramet 'dfa' ja 'dfb' yhdeksi DataFrameksi 'df' sarakkeen 'ID' perusteella,
* jos sarakkeessa 'Ft' on merkkijono 'DIESEL', korvataan se merkkijonolla 'Diesel',
* jos sarakkeessa 'Ft' on merkkijono 'PETROL', korvataan se merkkijonolla 'Petrol',
* tekee uuden DataFramen, jossa on laskettu moottorien keskiarvotilavuudet (sarake 'ec (cm3)') ryhmiteltynä sarakkeiden 'Mh' ja 'Ft' mukaan,
* esittää moottorien keskiarvotilavuudet kokonaislukujen tarkkuudella,
* muodostaa monitasoisen DataFramen, jossa sarakeindekseinä ovat monikot ('ec (cm3)',   'Diesel') ja ('ec (cm3)',   'Petrol'),
* jos DataFramessa esiintyy NaN arvoja, niin DataFramen soluun kirjoitetaan merkkijono "no value",
* tulostaa vastaukseen 10 ensimmäistä riviä.


Tässä tehtävässä:
* url_a1=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA1.csv
* url_a2=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA2.csv
* url_b1=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB1.csv
* url_b2=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB2.csv

Esimerkiksi, autot(url_a1, url_a2, url_b1, url_b2):

<style type="text/css">
	table.tableizer-table {
		font-size: 12px;
		border: 1px solid #CCC; 
		font-family: Arial, Helvetica, sans-serif;
	} 
	.tableizer-table td {
		padding: 4px;
		margin: 3px;
		border: 1px solid #CCC;
	}
	.tableizer-table th {
		background-color: #104E8B; 
		color: #FFF;
		font-weight: bold;
	}
</style>
<table class="tableizer-table">
<thead><tr class="tableizer-firstrow"><th></th><th>ec (cm3)          </th><th>&nbsp;</th></tr>
 <tr><td>Ft                   </td><td>  Diesel </td><td> Petrol</td></tr>
 <tr><td>Mh                   </td><td>         </td><td>       </td></tr>
 </thead><tbody>
 <tr><td>AA-IVA               </td><td>1995</td><td>3596</td></tr>
 <tr><td>AUTOMOBILES CITROEN  </td><td>1560</td><td>1131</td></tr>
 <tr><td>FORD WERKE GMBH      </td><td>1615</td><td>1292</td></tr>
 <tr><td>KIA                  </td><td>no value </td><td>1580</td></tr>
 <tr><td>NISSAN               </td><td>1559</td><td>1248</td></tr>
 <tr><td>SKODA                </td><td>1800</td><td>1191</td></tr>
 <tr><td>TESLA                </td><td>no value </td><td>no value</td></tr>
</tbody></table>

In [135]:
def autot(url_a1, url_a2, url_b1, url_b2):
    import pandas as pd
    import numpy as np
    df_a1 = pd.read_csv(url_a1, sep=';')
    df_a2 = pd.read_csv(url_a2, sep=';')
    df_b1 = pd.read_csv(url_b1, sep=';')
    df_b2 = pd.read_csv(url_b2, sep=';')    
        
    dfa = pd.concat([df_a1, df_a2], ignore_index = True)
    dfb = pd.concat([df_b1, df_b2], ignore_index = True)
            
    df = pd.merge(dfa, dfb, on = 'ID')
    
    df['Ft'] = df['Ft'].str.replace('DIESEL', 'Diesel')
    df['Ft'] = df['Ft'].str.replace('PETROL', 'Petrol')
      
    
    df3 = df.pivot_table(['ec (cm3)'], index = ['Mh'], columns = ['Ft'], dropna = False).applymap("{:.0f}".format)
    df3 = df3[[('ec (cm3)','Diesel'), ('ec (cm3)','Petrol')]]
    df3[('ec (cm3)','Diesel')].replace('nan', 'no value', inplace=True)
    df3[('ec (cm3)','Petrol')].replace('nan', 'no value', inplace=True)
    tulos = df3.head(10)
    # Älä kirjoita muita print komentoja ennen tulos muuttujan tulostusta.
    print(tulos)
    
# Testaa funktiotasi:
url_a1='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA1.csv'
url_a2='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA2.csv'
url_b1='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB1.csv'
url_b2='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB2.csv'

autot(url_a1, url_a2, url_b1, url_b2)

                     ec (cm3)          
Ft                     Diesel    Petrol
Mh                                     
AA-IVA                   1995      3596
AUTOMOBILES CITROEN      1560      1131
FORD WERKE GMBH          1615      1292
KIA                  no value      1580
NISSAN                   1559      1248
SKODA                    1800      1191
TESLA                no value  no value


### 5. Autodatan yhdistäminen ja ryhmittely monitasoiseksi DataFrameksi (2)

Tee funktio autot(url_a1, url_a2, url_b1, url_b2), joka tekee järjestyksessä seuraavaa:
* lataa datat DataFrameihin osoitteista (url_a1, url_a2, url_b1, url_b2), joissa erotinmerkkinä on puolipiste (';'),
* yhdistää uuteen DataFrameen 'dfa' autoihin liittyvät datat url_a1:sta ja url_a2:sta (tutki tarvittaessa mitä dataa urleista on saatavilla),
* yhdistää uuteen DataFrameen 'dfb' autoihin liittyvät datat url_b1:sta ja url_b2:sta (tutki tarvittaessa mitä dataa urleista on saatavilla),
* yhdistää DataFramet 'dfa' ja 'dfb' yhdeksi DataFrameksi 'df' sarakkeen 'ID' perusteella,
* jos sarakkeessa 'Ft' on merkkijono 'DIESEL', korvataan se merkkijonolla 'Diesel',
* jos sarakkeessa 'Ft' on merkkijono 'PETROL', korvataan se merkkijonolla 'Petrol',
* tallentaa uuteen DataFrameen sarakkeet 'Enedc (g/km)', 'MS' ja 'Ft',
* korvaa sarakkeen nimen 'Enedc (g/km)' nimellä 'Avg emissions (g/km)', 
* korvaa sarakkeen nimen 'MS' nimellä 'Member State', 
* korvaa sarakkeen nimen 'Ft' nimellä 'Fuel type', 
* laskee uuteen DataFrameen keskimääräiset päästöt (sarake 'Avg emissions (g/km)') maittain (sarake 'Member State') ja polttoainetyypeittäin (sarake 'Fuel type'),
* jättää edellä tuotettuun monitasoiseen DataFrameen vain 'Diesel' ja 'Petrol' sarakkeet,
* laskee uuteen sarakkeeseen 'D-P' kuinka paljon sarkkeen 'Diesel' arvo on suurempi kuin sarakkeen 'Petrol',
* järjestää DataFramen sarakkeen 'D-P' mukaan laskevaan järjestykseen,
* pyöristää ja esittää DataFramen lukuarvot kokonaislukujen tarkkuudella,
* tulostaa vastaukseen DataFramesta viisi ensimmäistä riviä.

Tässä tehtävässä:
* url_a1=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA1.csv
* url_a2=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA2.csv
* url_b1=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB1.csv
* url_b2=https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB2.csv

Esimerkiksi, autot(url_a1, url_a2, url_b1, url_b2):

<style type="text/css">
	table.tableizer-table {
		font-size: 12px;
		border: 1px solid #CCC; 
		font-family: Arial, Helvetica, sans-serif;
	} 
	.tableizer-table td {
		padding: 4px;
		margin: 3px;
		border: 1px solid #CCC;
	}
	.tableizer-table th {
		background-color: #104E8B; 
		color: #FFF;
		font-weight: bold;
	}
</style>
<table class="tableizer-table">
<thead><tr class="tableizer-firstrow"><th></th><th>Avg emissions (g/km)</th><th>&nbsp;</th><th>&nbsp;</th><th>&nbsp;</th></tr>
 <tr><td>Fuel type</td><td>&nbsp;</td><td>Diesel</td><td>Petrol</td><td>D-P</td></tr>
 <tr><td>Member State</td><td>&nbsp;</td><td>&nbsp;</td><td>&nbsp;</td><td>&nbsp;</td></tr>
    </thead><tbody>
 <tr><td>EE</td><td>&nbsp;</td><td>132</td><td>108</td><td>25</td></tr>
 <tr><td>FI</td><td>&nbsp;</td><td>137</td><td>112</td><td>25</td></tr>
 <tr><td>LV</td><td>&nbsp;</td><td>131</td><td>108</td><td>23</td></tr>
 <tr><td>GR</td><td>&nbsp;</td><td>122</td><td>100</td><td>22</td></tr>
 <tr><td>CZ</td><td>&nbsp;</td><td>122</td><td>110</td><td>12</td></tr>
</tbody></table>

In [177]:
def autot(url_a1, url_a2, url_b1, url_b2):
    import pandas as pd
    import numpy as np
    df_a1 = pd.read_csv(url_a1, sep=';')
    df_a2 = pd.read_csv(url_a2, sep=';')
    df_b1 = pd.read_csv(url_b1, sep=';')
    df_b2 = pd.read_csv(url_b2, sep=';')    
        
    dfa = pd.concat([df_a1, df_a2], ignore_index = True)
    dfb = pd.concat([df_b1, df_b2], ignore_index = True)
            
    df = pd.merge(dfa, dfb, on = 'ID')
    
    df['Ft'] = df['Ft'].str.replace('DIESEL', 'Diesel')
    df['Ft'] = df['Ft'].str.replace('PETROL', 'Petrol')
    
    df2 = df[['Enedc (g/km)', 'MS', 'Ft']].copy()
    df2.rename(columns={"Enedc (g/km)": "Avg emissions (g/km)", "MS": "Member State", "Ft": "Fuel type"}, inplace = True)
    
    df3 = df2.pivot_table(['Avg emissions (g/km)'], index = ['Member State'], columns = ['Fuel type'], dropna = False).astype(float)
    df3 = df3[[('Avg emissions (g/km)','Diesel'), ('Avg emissions (g/km)','Petrol')]]
    
    df3[('Avg emissions (g/km)','D-P')] = df3[('Avg emissions (g/km)','Diesel')] - df3[('Avg emissions (g/km)','Petrol')]
    df3 = df3.sort_values(('Avg emissions (g/km)','D-P'), ascending = False)
    df3 = df3.applymap("{:.0f}".format)
    tulos = df3.head(5)
    # Älä kirjoita muita print komentoja ennen tulos muuttujan tulostusta.
    print(tulos)
    
# Testaa funktiotasi:
url_a1='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA1.csv'
url_a2='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotA2.csv'
url_b1='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB1.csv'
url_b2='https://student.labranet.jamk.fi/~huoptu/datananal_s20/kerta4/harjoitukset/autotB2.csv'

autot(url_a1, url_a2, url_b1, url_b2)

             Avg emissions (g/km)           
Fuel type                  Diesel Petrol D-P
Member State                                
EE                            132    108  25
FI                            137    112  25
LV                            131    108  23
GR                            122    100  22
CZ                            122    110  12
